### DISCUSSION TITLE
What is the simplest way of adding wind through xml script ?

### DISCUSSION THREAD

**Discussion by rishi1906** [JSBSim-Team/jsbsim#1099] [Q&A]
I am looking for a way to add wind in one direction (say top to down) varying with height. I went through threads where similar thing has been discussed, but couldn't find a quick/easy way to include wind through script. @seanmcleod @bcoconni could you help me with this.

Note : I am using Ardupilot SITL for simulation, it allows to set SIM_WIND_SPD parameter, I could see the effect as well in mission planner. I am not sure whether this allows JSBSim to get info of wind or not.

**Comment by seanmcleod** [comment]
Okay, so you want to vary the wind with altitude, however the wind parameters don't take altitude as an argument.

https://github.com/JSBSim-Team/jsbsim/blob/c867cc7116283f08850db36444baf2577b967b67/src/models/atmosphere/FGWinds.cpp#L497-L506

So you would need to have an external process that monitored the aircraft's altitude and then updated the wind as the altitude changed via the available wind properties.

If you're willing to modify the aircraft's FDM file you could potentially add a function which used the aircraft's altitude multiplied by some factor to update the wind property.

**Reply by rishi1906** [reply]
Thanks for replying, I am not sure how to assign values to these parameters, there are so many variables under atmosphere/.. how do I decide which one is to be set. 

Could you possibly suggest me if effect of wind could be modeled where flight dynamics equations are coded?

**Comment by seanmcleod** [comment]
This is what I was suggesting if you're happy to modify each aircraft's FDM.

```xml
<aerodynamics>
   <function name="aero/function/windcalc" copyto="atmosphere/wind-mag-fps">
      <description>Varying wind with altitude</description>
      <table>
         <independentVar>position/h-sl-ft</independentVar>
         <tableData>
                0        0
               40000    40
         </tableData>
       </table>
   </function>
```

**Reply by rishi1906** [reply]
Yes, I am open to modifying aircraft's FDM. Is this a dummy code or will it work as it is?

**Reply by seanmcleod** [reply]
Not dummy code, I added that as is to the 737's FDM and then ran the `737_cruise.xml` script and added a property notification for `atmosphere/wind-mag-fps` in the script:

```xml
		<event name="Repeating Notify" persistent="true">
			<description>Output message at 5 second intervals</description>
			<notify>
				<property>propulsion/engine[0]/n2</property>
				<property>propulsion/engine[1]/n2</property>
				<property>propulsion/engine[0]/thrust-lbs</property>
				<property>propulsion/engine[1]/thrust-lbs</property>
				<property apply="convert-lbs-To-N" caption="propulsion/engine[0]/thrust-N">propulsion/engine[0]/thrust-lbs</property>
				<property apply="convert-lbs-To-N" caption="propulsion/engine[1]/thrust-N">propulsion/engine[1]/thrust-lbs</property>
				<property>position/h-agl-ft</property>
				<property>velocities/vc-kts</property>
				<property>velocities/vc-fps</property>
				<property>velocities/vt-fps</property>
				<property>attitude/phi-rad</property>
				<property>attitude/theta-rad</property>
				<property>attitude/psi-rad</property>
				<property>inertia/weight-lbs</property>
				<property>atmosphere/wind-mag-fps</property>
			</notify>
			<condition> simulation/sim-time-sec >= simulation/notify-time-trigger </condition>
			<set name="simulation/notify-time-trigger" value="5" type="FG_DELTA"/>
		</event>
```

And confirmed that the wind magnitude was being set. I did drop the maximum wind at 40,000ft to 20, otherwise the aircraft couldn't be trimmed.

```sh
Repeating Notify (Event 2) executed at time: 100.0
    propulsion/engine[0]/n2 = 97.4
    propulsion/engine[1]/n2 = 97.4
    propulsion/engine[0]/thrust-lbs = 7129.4
    propulsion/engine[1]/thrust-lbs = 7129.4
    propulsion/engine[0]/thrust-N = 31713.2
    propulsion/engine[1]/thrust-N = 31713.2
    position/h-agl-ft = 29969.6
    velocities/vc-kts = 291.0
    velocities/vc-fps = 491.2
    velocities/vt-fps = 764.0
    attitude/phi-rad = -0.0
    attitude/theta-rad = 0.0
    attitude/psi-rad = 3.9
    inertia/weight-lbs = 107000.0
    atmosphere/wind-mag-fps = 15.0
```

**Reply by rishi1906** [reply]
I need to know whether adding 


```
>    <function name="aero/function/windcalc" copyto="atmosphere/wind-mag-fps">
>       <description>Varying wind with altitude</description>
>       <table>
>          <independentVar>position/h-sl-ft</independentVar>
>          <tableData>
>                 0        0
>                40000    40
>          </tableData>
>        </table>
>    </function>

```
in `<aerodynamics>` tag in aircraft's FDM will add wind effect. If so where in FDM equation this effect has been captured in JSBSim? 

I am not sure how this function` "aero/function/windcalc"` will get invoked by JSBSim on its own while doing simulation with Ardupilot SITL?
 
You seem to have used `737_cruise.xml `script to invoke it, could you correct me if I am wrong?

**Reply by seanmcleod** [reply]
> I need to know whether adding in <aerodynamics> tag in aircraft's FDM will add wind effect. 

Well, why not run the 737 cruise script with and without the wind and compare the difference?

As I posted above, here is the result with wind.

```sh
Repeating Notify (Event 2) executed at time: 100.0
    propulsion/engine[0]/n2 = 97.4
    propulsion/engine[1]/n2 = 97.4
    propulsion/engine[0]/thrust-lbs = 7129.4
    propulsion/engine[1]/thrust-lbs = 7129.4
    propulsion/engine[0]/thrust-N = 31713.2
    propulsion/engine[1]/thrust-N = 31713.2
    position/h-agl-ft = 29969.6
    velocities/vc-kts = 291.0
    velocities/vc-fps = 491.2
    velocities/vt-fps = 764.0
    attitude/phi-rad = -0.0
    attitude/theta-rad = 0.0
    attitude/psi-rad = 3.9
    inertia/weight-lbs = 107000.0
    atmosphere/wind-mag-fps = 15.0
```

And the result with 0 wind.

```sh
Repeating Notify (Event 2) executed at time: 100.0
    propulsion/engine[0]/n2 = 97.2
    propulsion/engine[1]/n2 = 97.2
    propulsion/engine[0]/thrust-lbs = 7023.8
    propulsion/engine[1]/thrust-lbs = 7023.8
    propulsion/engine[0]/thrust-N = 31243.4
    propulsion/engine[1]/thrust-N = 31243.4
    position/h-agl-ft = 30006.4
    velocities/vc-kts = 285.0
    velocities/vc-fps = 481.1
    velocities/vt-fps = 750.0
    attitude/phi-rad = -0.0
    attitude/theta-rad = 0.0
    attitude/psi-rad = 3.9
    inertia/weight-lbs = 107000.0
    atmosphere/wind-mag-fps = 0.0
```

> I am not sure how this function "aero/function/windcalc" will get invoked by JSBSim on its own while doing simulation with Ardupilot SITL?

On every time step all the functions in the `aerodynamics` section will be evaluated by JSBSim.

> You seem to have used 737_cruise.xml script to invoke it, could you correct me if I am wrong?

Yes I use that script file to test, but there is nothing unique or special about that specific script file.

> If so where in FDM equation this effect has been captured in JSBSim?

https://github.com/JSBSim-Team/jsbsim/blob/c867cc7116283f08850db36444baf2577b967b67/src/models/FGAuxiliary.cpp#L144-L146

**Reply by rishi1906** [reply]
![image](https://github.com/JSBSim-Team/jsbsim/assets/16606630/9563a0d4-0f72-4b25-a675-0de4cdb334ec)
Even after adding the linear profile of wind wrt to height in aircraft's FDM, its not being updated, it shows zero value for wind parameters as shown above.

**Reply by rishi1906** [reply]
```
<function name="aero/function/windcalc" copyto="atmosphere/wind-mag-fps">
        <description>Varying wind with altitude</description>
        <table>
            <independentVar>position/h-sl-ft</independentVar>
            <tableData>
                0.0     0.0
                2.0     0.0
                32.81	5.00
                62.34	6.50
                91.86	8.00
                121.39	9.50
                150.92	11.00
                180.45	12.50
                209.97	14.00
                239.50	15.50
                269.03	17.00
                298.56	18.50
                328.08	20.00
            </tableData>
        </table>
        </function>

```

I added above lines under ` <aerodynamics> `tag.

**Reply by rishi1906** [reply]
Got it working! Thanks! I had put wind speed in mps. converted it to fps and it worked. 
![Screenshot from 2024-07-06 19-37-27](https://github.com/JSBSim-Team/jsbsim/assets/16606630/c7bbd2de-5b6c-495d-ac78-2395793e49bc)

**Comment by seanmcleod** [comment]
I've opened a new discussion - https://github.com/JSBSim-Team/jsbsim/discussions/1108 about adding support for functions in script files.
